pkg/await/apps_statefulset.go

```diff
@@ -0,0 +1,451 @@
+package await
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-kubernetes/pkg/client"
+	"github.com/pulumi/pulumi-kubernetes/pkg/openapi"
+	"github.com/pulumi/pulumi/pkg/diag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+)
+
+// ------------------------------------------------------------------------------------------------
+
+// Await logic for apps/v1beta1/StatefulSet, apps/v1beta2/StatefulSet,
+// and apps/v1/StatefulSet.
+//
+// The goal of this code is to provide a fine-grained account of the status of a Kubernetes
+// StatefulSet as it is being initialized. The idea is that if something goes wrong early, we want to
+// alert the user so they can cancel the operation instead of waiting for timeout (~10 minutes).
+//
+// A StatefulSet is a construct that allows users to specify how to execute an update to a stateful
+// application that is replicated some number of times in a cluster. When an application is updated,
+// the StatefulSet will incrementally roll out the new version (according to the policy requested by
+// the user). When the new application Pods becomes "live" (as specified by the liveness and
+// readiness probes), the old Pods are killed and deleted.
+//
+// The success conditions are somewhat complex:
+//
+//   1. `.status.replicas`, `.status.currentReplicas` and `.status.readyReplicas` match the
```

Comment:
Ah, looks like we also have to check `.status.updatedReplicas` if it's an update rather than a create?

Reply:
I'm using it for the status messages, but it's not actually part of the success condition.

Here's a table that's illustrative (I'll add to the doc as well):
```
Current replicas    Ready replicas  Updated replicas    Notes
3                   3               --
<Update image>
2                   3               --                  observedGeneration/updateRevision changes
2                   2               --
2                   2               1
1                   3               1
1                   2               1
1                   2               2
--                  3               2
--                  2               2
--                  2               3
3                   3               3                   currentRevision updated
3                   3               --
```